src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -814,9 +814,19 @@ ExecuteSingleModifyTask(CitusScanState *scanState, Task *task, bool multipleTask
 		LockPartitionRelations(relationId, RowExclusiveLock);
 	}
 
-	if (task->taskType == MODIFY_TASK)
+	/*
+	 * Prevent replicas of the same shard from diverging. We don't
+	 * need to acquire lock for TRUNCATE_TASK since it already
+	 * acquires AccessExclusiveLock on the relation, and blocks any
+	 * concurrent operation.
+	 *
+	 * Note that master_modify_multiple_shards() does not set
+	 * the operation type, thus does not acquire the executor
+	 * shard locks, which is a known issue.
```

Comment:
There seems to be less harm it taking a lock on `TRUNCATE` then not taking a lock on `master_modify_multiple_shards`

Reply:
I wanted to keep the behavior not changed because `master_modify_multiple_shards()` doesn't do `AcquireExecutorShardLock ()`. I thought that's an intended behavior. 

Do we prefer to fix it now?

Reply:
It still goes through `AcquireExecutorMultiShardLocks`, e.g. UPDATEs on RF>1 take exclusive locks.

Reply:
That's true, I missed that by adding a break point to `AcquireExecutorShardLock()` and not considering `AcquireExecutorMultiShardLocks()` at all. I should have stopped at `LockShardResource()`.

The minor difference between `AcquireExecutorShardLock ()` and `AcquireExecutorMultiShardLocks ()` is that the former expects `commandType` where `master_modify_multiple_shards` doesn't currently provide when calling `ExecuteSingleModifyTask()`.

We could either explicitly pass `commandType`  to `ExecuteSingleModifyTask()` or leave as is. Are you OK with passing `commandType` to `ExecuteSingleModifyTask` along with `Task`? I'm a bit hesitant doing that since we already have `TaskType` which is unfortuneatly not sufficient for this purpose :/

(I'm avoiding the alternative to add the commandType to task, which sound not a good idea.)

Reply:
I think adding a parameter is probably ok. It'd be good to gravitate towards taking scanState out of the API, since there might be significant API changes in PG12.